cli/doc/printer.rs

```diff
@@ -12,540 +12,449 @@
-  if !params.is_empty() {
-    for param in params {
-      rendered += param.name.as_str();
-      if param.optional {
-        rendered += "?";
+  fn format_signature(
+    &self,
+    w: &mut Formatter<'_>,
+    node: &doc::DocNode,
+    indent: i64,
+  ) -> FmtResult {
+    match node.kind {
+      DocNodeKind::Function => self.format_function_signature(w, node, indent),
+      DocNodeKind::Variable => self.format_variable_signature(w, node, indent),
+      DocNodeKind::Class => self.format_class_signature(w, node, indent),
+      DocNodeKind::Enum => self.format_enum_signature(w, node, indent),
+      DocNodeKind::Interface => {
+        self.format_interface_signature(w, node, indent)
+      }
+      DocNodeKind::TypeAlias => {
+        self.format_type_alias_signature(w, node, indent)
       }
-      if let Some(ts_type) = param.ts_type {
-        rendered += ": ";
-        rendered += render_ts_type(ts_type).as_str();
+      DocNodeKind::Namespace => {
+        self.format_namespace_signature(w, node, indent)
       }
-      rendered += ", ";
     }
-    rendered.truncate(rendered.len() - 2);
   }
-  rendered
-}
 
-fn render_ts_type(ts_type: doc::ts_type::TsTypeDef) -> String {
-  if ts_type.kind.is_none() {
-    return "<UNIMPLEMENTED>".to_string();
-  }
-  let kind = ts_type.kind.unwrap();
-  match kind {
-    TsTypeDefKind::Array => {
-      format!("{}[]", render_ts_type(*ts_type.array.unwrap()))
-    }
-    TsTypeDefKind::Conditional => {
-      let conditional = ts_type.conditional_type.unwrap();
-      format!(
-        "{} extends {} ? {} : {}",
-        render_ts_type(*conditional.check_type),
-        render_ts_type(*conditional.extends_type),
-        render_ts_type(*conditional.true_type),
-        render_ts_type(*conditional.false_type)
-      )
-    }
-    TsTypeDefKind::FnOrConstructor => {
-      let fn_or_constructor = ts_type.fn_or_constructor.unwrap();
-      format!(
-        "{}({}) => {}",
-        if fn_or_constructor.constructor {
-          "new "
-        } else {
-          ""
-        },
-        render_params(fn_or_constructor.params),
-        render_ts_type(fn_or_constructor.ts_type),
-      )
-    }
-    TsTypeDefKind::IndexedAccess => {
-      let indexed_access = ts_type.indexed_access.unwrap();
-      format!(
-        "{}[{}]",
-        render_ts_type(*indexed_access.obj_type),
-        render_ts_type(*indexed_access.index_type)
-      )
-    }
-    TsTypeDefKind::Intersection => {
-      let intersection = ts_type.intersection.unwrap();
-      let mut output = "".to_string();
-      if !intersection.is_empty() {
-        for ts_type in intersection {
-          output += render_ts_type(ts_type).as_str();
-          output += " & "
-        }
-        output.truncate(output.len() - 3);
+  // TODO(SyrupThinker) this should use a JSDoc parser
+  fn format_jsdoc(
+    &self,
+    w: &mut Formatter<'_>,
+    jsdoc: &str,
+    indent: i64,
+    details: bool,
+  ) -> FmtResult {
```

Comment:
Does this method change current summary output?

Reply:
Yes.
Previously it printed the entire documentation and removed newlines assuming that it is Markdown formatted.
IMO that made it unusable because either long documentation comments drowned out the relevant information or they were hard to read.

Now it doesn't remove newlines making the output more readable and in summaries it stops at the first blank line.
Usually doc comments are written with a quick summary at the beginning so I consider it an acceptable assumption.
They are also likely formatted in a way that is readable in a text editor so that should be usable in the terminal too.

As denoted in a comment it should use an actual JSDoc parser once one is available to utilize the `@summary` tag and take other tags into account.

Reply:
I see, thanks for explanation. Does it print full JSdoc in detail mode? (`deno doc ./myfile.ts SomeClass`)

> As denoted in a comment it should use an actual JSDoc parser once one is available to utilize the `@summary` tag and take other tags into account.

Sure, waiting on https://github.com/swc-project/swc/pull/829

Reply:
It does:
![image](https://user-images.githubusercontent.com/7753242/86410044-9bf15200-bca9-11ea-8529-c0525b707695.png)